sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -66,980 +85,305 @@ case class ScalaUDF(
+      s"this.$inputExpressionEncoderTerm = ($expressionEncoderClassName)$rowEncoderClassName" +
+        s".apply((($scalaUDFClassName)expressions" +
+          s"[$scalaUDFTermIdx]).getInputSchema());")
+
+    // Generate code for output encoder
+    val outputExpressionEncoderTerm = ctx.freshName("outputExpressionEncoder")
+    ctx.addMutableState(expressionEncoderClassName, outputExpressionEncoderTerm,
+      s"this.$outputExpressionEncoderTerm = ($expressionEncoderClassName)$rowEncoderClassName" +
+        s".apply((($scalaUDFClassName)expressions[$scalaUDFTermIdx]).getDataType());")
 
     val resultTerm = ctx.freshName("result")
 
-    // This must be called before children expressions' codegen
-    // because ctx.references is used in genCodeForConverter
-    val converterTerms = (0 until children.size).map(genCodeForConverter(ctx, _))
-
     // Initialize user-defined function
     val funcClassName = s"scala.Function${children.size}"
 
     val funcTerm = ctx.freshName("udf")
-    val funcExpressionIdx = ctx.references.size - 1
     ctx.addMutableState(funcClassName, funcTerm,
       s"this.$funcTerm = ($funcClassName)((($scalaUDFClassName)expressions" +
-        s"[$funcExpressionIdx]).userDefinedFunc());")
+        s"[$scalaUDFTermIdx]).userDefinedFunc());")
 
     // codegen for children expressions
     val evals = children.map(_.gen(ctx))
+    val evalsArgs = evals.map(_.value).mkString(", ")
+    val evalsAsSeq = s"$javaConversionClassName.asScalaIterable" +
+      s"(java.util.Arrays.asList($evalsArgs)).toList()"
+    val inputInternalRowTerm = ctx.freshName("inputRow")
+    val inputInternalRow = s"$rowClass $inputInternalRowTerm = " +
+      s"($rowClass)$inputExpressionEncoderTerm.fromRow(InternalRow.fromSeq($evalsAsSeq));"
 
     // Generate the codes for expressions and calling user-defined function
     // We need to get the boxedType of dataType's javaType here. Because for the dataType
     // such as IntegerType, its javaType is `int` and the returned type of user-defined
     // function is Object. Trying to convert an Object to `int` will cause casting exception.
     val evalCode = evals.map(_.code).mkString
-    val funcArguments = converterTerms.zip(evals).map {
-      case (converter, eval) => s"$converter.apply(${eval.value})"
-    }.mkString(",")
+
+    val funcArguments = (0 until children.size).map { i =>
+      s"$inputInternalRowTerm.get($i)"
+    }.mkString(", ")
+
+    val rowParametersTerm = ctx.freshName("rowParameters")
+    val innerRow = s"$rowClass $rowParametersTerm = $rowClassName.apply(" +
+      s"$javaConversionClassName.asScalaIterable" +
+      s"(java.util.Arrays.asList($funcTerm.apply($funcArguments))).toList());"
+    val internalRowTerm = ctx.freshName("internalRow")
+    val internalRow = s"$internalRowClassName $internalRowTerm = ($internalRowClassName)" +
+      s"${outputExpressionEncoderTerm}.toRow($rowParametersTerm).copy();"
```

Comment:
Hmm, basically Encoder does in `toRow` is to project input row to internal row using `GenerateUnsafeProjection`. I think we can't just embed the generated code here. How do you think?